File: LynxExample/android/app/src/main/java/com/lynxscreens/elements/LynxColorBoxComponent.kt
package com.lynxscreens.elements

import android.content.Context
import android.graphics.Color
import com.lynx.tasm.behavior.LynxContext
import com.lynx.tasm.behavior.LynxProp
import com.lynx.tasm.behavior.ui.UIGroup
import com.lynx.tasm.behavior.ui.view.AndroidView

class LynxColorBoxComponent(context: LynxContext) : UIGroup<AndroidView>(context) {
    override fun createView(context: Context?): AndroidView = AndroidView(context)

    // Note: This doesn't seem to be good place to apply updates (considering Screens impl), but
    // it's sufficient for basic testing and demonstration purposes.
    override fun onLayoutUpdated() {
        super.onLayoutUpdated()

        // Retrieve the corresponding ShadowNode from the Lynx context by node sign (unique ID)
        lynxContext.findShadowNodeBySign(sign)?.let {
            // Ensure that the retrieved node is an instance of our custom ShadowNode class
            if (it is LynxColorBoxShadowNode) {
                // Store the previously set size
                val oldWidth = this.width
                val oldHeight = this.height

                // Perform some calculations for updating the size
                val (newWidth, newHeight) = adjustViewSize()

                // If the size has changed, update the ShadowNode
                if (oldWidth != newWidth || oldHeight != newHeight) {
                    it.updateSize(newWidth, newHeight)
                }
            }
        }
    }

    // Returns a hardcoded width and height for testing purposes
    private fun adjustViewSize(): Pair<Int, Int> {
        val sizeDp = 300

        val density = mContext.resources.displayMetrics.density

        val widthPx = (sizeDp * density).toInt()
        val heightPx = (sizeDp * density).toInt()

        return Pair(widthPx, heightPx)
    }

    @LynxProp(name = "backgroundColorHex")
    fun setBackgroundColorHex(value: String) {
        try {
            val color = Color.parseColor(value)
            mView.setBackgroundColor(color)
        } catch (e: IllegalArgumentException) {
        }
    }
}

File: LynxExample/ios/LynxScreens/elements/LynxColorBoxComponent.m
#import "LynxColorBoxComponent.h"
#import "LynxColorBoxShadowNode.h"

#import <Lynx/LynxComponentRegistry.h>
#import <Lynx/LynxPropsProcessor.h>
#import <Lynx/LynxShadowNodeOwner.h>

@implementation LynxColorBoxComponent

LYNX_LAZY_REGISTER_UI("color-box-view")

// Note: This doesn't seem to be good place to apply updates, but
// it's sufficient for basic testing and demonstration purposes.
- (void)layoutDidFinished {
    // Retrieve the corresponding ShadowNode from the Lynx context by node sign (unique ID)
    LynxColorBoxShadowNode *node = (LynxColorBoxShadowNode*)[self.context.nodeOwner nodeWithSign:self.sign];
    // Ensure that the retrieved node is an instance of our custom ShadowNode class
    if ([node isKindOfClass:LynxColorBoxShadowNode.class]) {
        // Store the previously set size
        CGSize preSize = node.uiSize;
        
        // Perform some calculations for updating the size
        CGSize updatedSize = [self adjustViewSize];
        
        // If the size has changed, update the ShadowNode and request a re-layout
        if (!CGSizeEqualToSize(preSize, updatedSize)) {
            node.uiSize = updatedSize;
            [node setNeedsLayout];
        }
    }
}

// Returns a hardcoded width and height for testing purposes
- (CGSize)adjustViewSize {
    return CGSizeMake(300, 300);
}

LYNX_PROP_SETTER("backgroundColorHex", setBackgroudColorHex, NSString *) {
    self.view.backgroundColorHex = value;
}

- (UIView *)createView {
    UIView *colorBoxView = [[LynxColorBoxView alloc] init];
    return colorBoxView;
}

@end
